ms-windows/build-tools.php

```diff
@@ -0,0 +1,1561 @@
+
+		@mkdir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/logs", 0775, true);
+
+		// Specialty recursive file copy to ignore certain file extensions in the build tree.
+		function CopyBuildFiles($srcdir, $destdir, $path, $recurse, $pattern = "/.*/")
+		{
+			$srcpath = $srcdir . "/" . $path;
+
+			$excludeexts = array(
+				".c" => true,
+				".d" => true,
+				".ec" => true,
+				".h" => true,
+				".obj" => true,
+				".ilk" => true,
+				".pdb" => true,  // While PDB files might be ideal for debug builds, the generated files are about 8 times larger than the DLLs and include local system and build path strings.
+				".in" => true,
+				".info" => true,
+				".md" => true,
+				".asm" => true,
+				".mar" => true,
+			);
+
+			if (!is_dir($srcpath))  return;
+
+			$dir = opendir($srcpath);
+			if ($dir)
+			{
+				while (($file = readdir($dir)) !== false)
+				{
+					if ($file !== "." && $file !== "..")
+					{
+						if (is_dir($srcpath . "/" . $file))
+						{
+							if ($recurse)  CopyBuildFiles($srcpath, $destdir . "/" . $path, $file, true, $pattern);
+						}
+						else if (strpos($file, ".") !== false && !isset($excludeexts[substr($file, strrpos($file, "."))]) && preg_match($pattern, $file))
+						{
+							@mkdir($destdir . "/" . $path, 0777, true);
+
+							copy($srcpath . "/" . $file, $destdir . "/" . $path . "/" . $file);
+						}
+					}
+				}
+
+				closedir($dir);
+			}
+		}
+
+		// Builds OpenSSL.
+		function RunBuild($ver, $arch, $baseenv, $buildnum, $pathmap, $archinfo, $configuretype, $buildtype, $outdir, $copymodes)
+		{
+			$echoprefix = "[" . $ver . "; " . $arch . "; Build " . $buildnum . "; " . $outdir . "]";
+
+			echo $echoprefix . " Started build at " . date("Y-m-d H:i:s") . ".\n";
+			$ts = time();
+
+			DeleteDirectory($pathmap["[[SOURCE_TEMP_DIR]]"]);
+
+			// Clone the source tree to the temporary architecture directory.
+			// Doing this allows multiple architectures for a base version to be built simultaneously.
+			echo $echoprefix . " Cloning source tree...\n";
+			CopyDirectory($pathmap["[[SOURCE_ORIG_DIR]]"], $pathmap["[[SOURCE_TEMP_DIR]]"]);
+
+
+			// Run perl Configure.
+			$cmd = escapeshellarg("perl.exe") . " Configure " . $archinfo["build"]["configure_target"] . " " . $configuretype . " " . $archinfo["build"]["configure_extra"];
+			echo $echoprefix . " Running '" . $cmd . "'...\n";
+
+			$errorfilename = $pathmap["[[BUILD_OUTPUT_DIR]]"] . "/logs/" . $buildnum . "_perl_configure_" . $outdir . "_error.log";
+			$result = RunCommand($cmd, $errorfilename, $pathmap["[[SOURCE_TEMP_DIR]]"], $baseenv);
+			if ($result === false)  CLI::DisplayError($echoprefix . " Failed to start process.");
+
+			file_put_contents($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/logs/" . $buildnum . "_perl_configure_" . $outdir . "_out.log", $result);
+
+			if (!file_exists($pathmap["[[SOURCE_TEMP_DIR]]"] . "/makefile"))  CLI::DisplayError($echoprefix . " Failed to generate 'makefile'.");
+
+
+			// Nothing is perfect.  Patch the makefile.
```

Comment:
The `info.json` files have a series of patches that I currently use.  I think the only one of note, based on your other comments, is the use of `/Zi` instead of `/Z7`.  `/Zi` produces library files that won't link across all versions of Visual Studio while `/Z7` tends to be more compatible when mixing versions of Visual Studio.

For the source release of OpenSSL, `/Zi` linkage is correct because the user (a developer) is unlikely to be mixing versions of VS.  For binary releases, `/Z7` is correct because the target VS version of the application developer is unknown and shipping potentially incompatible .pdb files is a pain.  You probably need both to be correct.  Patching is the simple, albeit hacky, solution.  The alternative would be to introduce a whole new series of supported targets specifically for binary production just to support a fairly minor change to the makefile.